include/sunmatrix/sunmatrix_kokkosdense.hpp

```diff
@@ -0,0 +1,405 @@
+/* -----------------------------------------------------------------------------
+ * Programmer(s): David J. Gardner @ LLNL
+ * -----------------------------------------------------------------------------
+ * SUNDIALS Copyright Start
+ * Copyright (c) 2002-2022, Lawrence Livermore National Security
+ * and Southern Methodist University.
+ * All rights reserved.
+ *
+ * See the top-level LICENSE and NOTICE files for details.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ * SUNDIALS Copyright End
+ * -----------------------------------------------------------------------------
+ * This is the header file for a dense SUNMarix implementation using Kokkos.
+ * ---------------------------------------------------------------------------*/
+
+#ifndef _SUNMATRIX_KOKKOSDENSE_HPP
+#define _SUNMATRIX_KOKKOSDENSE_HPP
+
+#include <KokkosBatched_Gemv_Decl.hpp>
+#include <KokkosBlas2_gemv.hpp>
+#include <Kokkos_Core.hpp>
+#include <nvector/nvector_kokkos.hpp>
+#include <sundials/sundials_base.hpp>
+#include <sundials/sundials_matrix.hpp>
+
+namespace sundials {
+namespace kokkos {
+
+// Forward declaration of Matrix class
+template<class ExecutionSpace = Kokkos::DefaultExecutionSpace,
+         class MemorySpace    = class ExecutionSpace::memory_space>
+class DenseMatrix;
+
+// Get the Kokkos dense matrix wrapped by a SUNMatrix
+template<class MatrixType>
+inline MatrixType* GetDenseMat(SUNMatrix A)
+{
+  return static_cast<MatrixType*>(A->content);
+}
+
+// =============================================================================
+// Everything in the implementation (impl) namespace is private and should not
+// be referred to directly in user code.
+// =============================================================================
+
+namespace impl {
+
+SUNMatrix_ID SUNMatGetID_KokkosDense(SUNMatrix A)
+{
+  return SUNMATRIX_KOKKOSDENSE;
+}
+
+template<class MatrixType>
+SUNMatrix SUNMatClone_KokkosDense(SUNMatrix A)
+{
+  auto A_mat{GetDenseMat<MatrixType>(A)};
+  auto new_mat{new MatrixType(A_mat->Blocks(), A_mat->BlockRows(),
```

Comment:
Is there a reason to use this constructor over the copy constructor?

Reply:
Originally the copy constructor was a deep copy. Now that it's a shallow copy, clone can just call the copy constructor.